arch/risc-v/src/mpfs/mpfs_corespi.c

```diff
@@ -0,0 +1,1498 @@
+
+#ifdef CONFIG_SPI_CS_CONTROL
+static void mpfs_spi_select(struct spi_dev_s *dev, uint32_t devid,
+                            bool selected)
+{
+  struct mpfs_spi_priv_s *priv = (struct mpfs_spi_priv_s *)dev;
+
+  priv->devid = devid & 0x1f;
+
+  if (selected)
+    {
+      modifyreg32(MPFS_SPI_SSELECT, 0, 1 << (priv->devid));
+    }
+  else
+    {
+      modifyreg32(MPFS_SPI_SSELECT, 1 << (priv->devid), 0);
+    }
+
+  spiinfo("devid: %u, CS: %s\n", devid, selected ? "select" : "free");
+}
+#endif
+
+/****************************************************************************
+ * Name: mpfs_spi_setfrequency
+ *
+ * Description:
+ *   Set the SPI frequency.
+ *
+ * Input Parameters:
+ *   dev -       Device-specific state data
+ *   frequency - The SPI frequency requested
+ *
+ * Returned Value:
+ *   Returns the current actual frequency selected
+ *
+ ****************************************************************************/
+
+static uint32_t mpfs_spi_setfrequency(struct spi_dev_s *dev,
+                                      uint32_t frequency)
+{
+  struct mpfs_spi_priv_s *priv = (struct mpfs_spi_priv_s *)dev;
+  uint32_t divider;
+
+  DEBUGASSERT(frequency > 0);
+
+  if (priv->enabled)
+    {
+      modifyreg32(MPFS_SPI_CONTROL, MPFS_SPI_ENABLE, 0);
+    }
+
+  priv->frequency = frequency;
+
+  /* Formula is SPICLK = PCLK/(2*(CFG_CLK + 1)) */
+
+  divider = ((MPFS_FPGA_PERIPHERAL_CLK / frequency) >> 1);
+  priv->actual = (uint32_t)frequency * divider;
+
+  DEBUGASSERT(divider >= 2u && divider <= 512u);
+
+  putreg32(divider, MPFS_SPI_CLK_GEN);
+
+  spiinfo("frequency=%u, actual=%u\n", priv->frequency, priv->actual);
+
+  if (priv->enabled)
+    {
+      modifyreg32(MPFS_SPI_CONTROL, 0, MPFS_SPI_ENABLE);
+    }
+
+  return priv->actual;
+}
+
+/****************************************************************************
+ * Name: mpfs_spi_setmode
+ *
+ * Description:
+ *   Set the SPI mode.
+ *
+ * Input Parameters:
+ *   dev -  Device-specific state data
+ *   mode - The SPI mode requested
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void mpfs_spi_setmode(struct spi_dev_s *dev, enum spi_mode_e mode)
+{
+  /* Mode is selected when the SPI block is being fabricated */
+
+  spierr("Changing SPI mode not supported\n");
+  DEBUGPANIC();
+}
+
+/****************************************************************************
+ * Name: mpfs_spi_setbits
+ *
+ * Description:
+ *   Set the number of bits per word.
+ *
+ * Input Parameters:
+ *   dev -  Device-specific state data
+ *   nbits - The number of bits in an SPI word.
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void mpfs_spi_setbits(struct spi_dev_s *dev, int nbits)
+{
+  /* Bitwidth is selected when the SPI block is being fabricated */
+
+  spierr("Changing SPI bitwidth not supported\n");
```

Comment:
Err only if setting to something else than the supported mode; I believe some of our SW calls this always with 8

Reply:
Ditto